src/main/java/org/opensearch/neuralsearch/query/BoundedLinearFeatureQuery.java

```diff
@@ -0,0 +1,230 @@
+                float score = boost * featureValue;
+                return Explanation.match(
+                    score,
+                    "Linear function on the " + fieldName + " field for the " + featureName + " feature, computed as w * S from:",
+                    Explanation.match(boost, "w, weight of this function"),
+                    Explanation.match(featureValue, "S, feature value")
+                );
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                Terms terms = Terms.getTerms(context.reader(), fieldName);
+                TermsEnum termsEnum = terms.iterator();
+                if (termsEnum.seekExact(new BytesRef(featureName)) == false) {
+                    return null;
+                }
+
+                final SimScorer scorer = new SimScorer() {
+                    @Override
+                    public float score(float freq, long norm) {
+                        return boost * decodeFeatureValue(freq);
+                    }
+                };
+                final ImpactsEnum impacts = termsEnum.impacts(PostingsEnum.FREQS);
+                final ImpactsDISI impactsDisi = new ImpactsDISI(impacts, impacts, scorer);
+
+                return new Scorer(this) {
+
+                    @Override
+                    public int docID() {
+                        return impacts.docID();
+                    }
+
+                    @Override
+                    public float score() throws IOException {
+                        return scorer.score(impacts.freq(), 1L);
+                    }
+
+                    @Override
+                    public DocIdSetIterator iterator() {
+                        return impactsDisi;
+                    }
+
+                    @Override
+                    public int advanceShallow(int target) throws IOException {
+                        return impactsDisi.advanceShallow(target);
+                    }
+
+                    @Override
+                    public float getMaxScore(int upTo) throws IOException {
+                        return impactsDisi.getMaxScore(upTo);
+                    }
+
+                    @Override
+                    public void setMinCompetitiveScore(float minScore) {
+                        impactsDisi.setMinCompetitiveScore(minScore);
+                    }
+                };
+            }
+        };
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        if (visitor.acceptField(fieldName)) {
+            visitor.visitLeaf(this);
+        }
+    }
+
+    @Override
+    public String toString(String field) {
+        return "BoundedLinearFeatureQuery(field=" + fieldName + ", feature=" + featureName + ", scoreUpperBound=" + scoreUpperBound + ")";
+    }
+
+    // the field and decodeFeatureValue are modified from FeatureField.decodeFeatureValue
+    static final int MAX_FREQ = Float.floatToIntBits(Float.MAX_VALUE) >>> 15;
```

Comment:
what is the value of this MAX_FREQ?

Reply:
This is copied from https://github.com/apache/lucene/blob/6d764c3397d00f93bd4273bd8d1c9e51d6e104e6/lucene/core/src/java/org/apache/lucene/document/FeatureField.java#L207, including the below method is from `FeatureField` but we changed it to make our new parameter scoreUpperBound work.